packages/ledgerstate/output.go

```diff
@@ -217,6 +261,213 @@ func OutputFromObjectStorage(key []byte, data []byte) (output objectstorage.Stor
+	sortedOutputs := make([]struct {
+		output           Output
+		outputSerialized []byte
+	}, 0)
+
+	// filter duplicates (store marshaled version so we don't need to marshal a second time during sort)
+	for _, output := range optionalOutputs {
+		marshaledOutput := output.Bytes()
+		marshaledOutputAsString := typeutils.BytesToString(marshaledOutput)
+
+		if _, seenAlready := seenOutputs[marshaledOutputAsString]; seenAlready {
+			continue
+		}
+		seenOutputs[marshaledOutputAsString] = types.Void
+
+		sortedOutputs = append(sortedOutputs, struct {
+			output           Output
+			outputSerialized []byte
+		}{output, marshaledOutput})
+	}
+
+	// sort outputs
+	sort.Slice(sortedOutputs, func(i, j int) bool {
+		return bytes.Compare(sortedOutputs[i].outputSerialized, sortedOutputs[j].outputSerialized) < 0
+	})
+
+	// create result
+	outputs = make(Outputs, len(sortedOutputs))
+	for i, sortedOutput := range sortedOutputs {
+		outputs[i] = sortedOutput.output
+	}
+
+	if len(outputs) < MinOutputCount {
+		panic(fmt.Sprintf("amount of Outputs (%d) failed to reach MinOutputCount (%d)", len(outputs), MinOutputCount))
+	}
+	if len(outputs) > MaxOutputCount {
+		panic(fmt.Sprintf("amount of Outputs (%d) exceeds MaxOutputCount (%d)", len(outputs), MaxOutputCount))
+	}
+
+	return
+}
+
+// OutputsFromBytes unmarshals a collection of Outputs from a sequence of bytes.
+func OutputsFromBytes(outputBytes []byte) (outputs Outputs, consumedBytes int, err error) {
+	marshalUtil := marshalutil.New(outputBytes)
+	if outputs, err = OutputsFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse Outputs from MarshalUtil: %w", err)
+		return
+	}
+	consumedBytes = marshalUtil.ReadOffset()
+
+	return
+}
+
+// OutputsFromMarshalUtil unmarshals a collection of Outputs using a MarshalUtil (for easier unmarshaling).
+func OutputsFromMarshalUtil(marshalUtil *marshalutil.MarshalUtil) (outputs Outputs, err error) {
+	outputsCount, err := marshalUtil.ReadUint16()
+	if err != nil {
+		err = xerrors.Errorf("failed to parse outputs count (%v): %w", err, cerrors.ErrParseBytesFailed)
+		return
+	}
+	if outputsCount < MinOutputCount {
+		err = xerrors.Errorf("amount of Outputs (%d) failed to reach MinOutputCount (%d): %w", outputsCount, MinOutputCount, cerrors.ErrParseBytesFailed)
+		return
+	}
+	if outputsCount > MaxOutputCount {
+		err = xerrors.Errorf("amount of Outputs (%d) exceeds MaxOutputCount (%d): %w", outputsCount, MaxOutputCount, cerrors.ErrParseBytesFailed)
+		return
+	}
+
+	var previousOutput Output
+	parsedOutputs := make([]Output, outputsCount)
+	for i := uint16(0); i < outputsCount; i++ {
+		if parsedOutputs[i], err = OutputFromMarshalUtil(marshalUtil); err != nil {
+			err = xerrors.Errorf("failed to parse Output from MarshalUtil: %w", err)
+			return
+		}
+
+		if previousOutput != nil && previousOutput.Compare(parsedOutputs[i]) != -1 {
+			err = xerrors.Errorf("order of Outputs is invalid: %w", cerrors.ErrParseBytesFailed)
```

Comment:
I discussed this with angelo and jonas and we decided to not have a separate "ErrSyntacticalValidationFailed" error because these things can also happen through things like bitflips on the transport layer and it essentially makes no sense to differentiate between Syntactical Errors and other Parsing error because any violation of the length or layout of the marshaled bytes is equally syntactically invalid as for example not sorting the outputs accordingly.